framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -285,104 +283,101 @@ protected boolean areEqualInHierarchy(
   }
 
-  private boolean isContainedWildcard(
+  /**
+   * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+   * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}. See
+   * {@link #isContainedBy(AnnotatedTypeMirror, AnnotatedTypeMirror, boolean)} for a full
+   * explanation.
+   *
+   * @param inside a possibly-contained type
+   * @param outsideLower the lower bound of the possibly-containing type
+   * @param outsideUpper a the upper bound of the possibly-containing type
+   * @param canBeCovariant whether or not type arguments are allowed to be covariant
+   * @return true if inside is contained by outside, or if canBeCovariant == true and {@code inside
+   *     <: outside}
+   */
+  protected boolean isContainedByBoundType(
       AnnotatedTypeMirror inside,
-      AnnotatedTypeMirror outside,
-      AnnotatedTypeMirror outsideUpperBound,
-      AnnotatedTypeMirror outsideLowerBound,
+      AnnotatedTypeMirror outsideLower,
+      AnnotatedTypeMirror outsideUpper,
       boolean canBeCovariant) {
-
-    if (inside.equals(outside)) {
-      // If they are equal, outside always contains inside.
-      return true;
-    }
-
-    if (inside.getKind() == TypeKind.WILDCARD) {
-      outsideUpperBound =
-          checker
-              .getTypeFactory()
-              .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-    }
-    while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-      if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-        return true;
+    try {
+      if (canBeCovariant) {
+        if (outsideLower.getKind() != TypeKind.NULL) {
+          return isSubtype(outsideLower, inside);
+        } else {
+          return isSubtype(inside, outsideUpper);
+        }
       }
-      outsideUpperBound = ((AnnotatedWildcardType) outsideUpperBound).getExtendsBound();
-    }
-
-    if (!TypesUtils.isErasedSubtype(
-        inside.getUnderlyingType(),
-        outsideUpperBound.getUnderlyingType(),
-        inside.atypeFactory.types)) {
-      // TODO: subtype is a wildcard that should have been captured, so just check the primary
-      // annotations.
-      AnnotationMirror subAnno = inside.getEffectiveAnnotationInHierarchy(currentTop);
-      AnnotationMirror superAnno = outsideUpperBound.getAnnotationInHierarchy(currentTop);
-      return qualifierHierarchy.isSubtype(subAnno, superAnno);
-    }
-
-    AnnotatedTypeMirror castedInside =
-        AnnotatedTypes.castedAsSuper(inside.atypeFactory, inside, outsideUpperBound);
-    if (!isSubtypeCaching(castedInside, outsideUpperBound)) {
-      return false;
-    }
-
-    if (outside.getKind() == TypeKind.WILDCARD && outsideLowerBound.getKind() == TypeKind.TYPEVAR) {
-      // tests/all-systems/Issue1991.java crashes without this.
-      return true;
+      return isSubtype(outsideLower, inside) && isSubtype(inside, outsideUpper);
```

Comment:
This looks suspicious to me.  Per the JLS algorithm, I think the first `isSubtype` call should be between outsideLower and the *bound* of inside, not inside itself.  (Likewise for the second `isSubtype` call.)

Reply:
If `inside` is a widlcard, then `isSubtype(outsideLower, inside)` calls `isSubtype(outsideLower, inside.getLowerBound())` and `isSubtype(inside, outsideUpper)` calls `isSubtype(inside.getUpperBound(), outsideUpper)`

Reply:
Since this is a variance from the algorithm as given in the JLS, I think it would be good to either:
 * change the code here (to slavishly copy the JLS algorithm) or 
 * note that the behavior is identical to the JLS algorithm.

That will avoid a similar question by readers in the future.  Thanks.